File: leetcode3/최원준/825. Friends Of Appropriate Ages.py
'''
1. 아이디어 :
- 나이가 120까지 있으니, 나이별로 묶어서 한번에 계산하는 방법

2. 시간복잡도 :
    O(n*n)

3. 자료구조/알고리즘 :
해시맵

'''

class Solution:
    def numFriendRequests(self, ages: List[int]) -> int:
        count = Counter(ages)
        ans = 0

        for x in count:
            for y in count:
                if self.can_send(x, y):
                    if x == y:
                        ans += count[x] * (count[x] - 1)
                    else:
                        ans += count[x] * count[y]

        return ans

    def can_send(self, x: int, y: int) -> bool:
        if y <= 0.5 * x + 7:
            return False
        if y > x:
            return False
        if y > 100 and x < 100:
            return False
        return True

File: leetcode3/이진희/825_Friends Of Appropriate Ages.java
/*

1. 아이디어 : 총 배열의 크기는 2만이지만, 나이는 120살까지 제한됨
            121크기의 배열을 만들어, 각 나이대별 인원을 구하고, 완전탐색으로 계산
            이때 서로 나이가 같을 경우 -> age[i] * (age[i]-1)
            서로 다를 경우 -> age[i] * age[j]

2. 시간복잡도 : O(N^2) -> O(14400)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public int numFriendRequests(int[] ages) {
        int[] age = new int[121];
        for (int a : ages) age[a]++;
        
        int ans = 0;

        for (int i = 1; i < 121; i++) {
            if (age[i] == 0) continue;
            for (int j = 1; j < 121; j++) {
                if (age[j] == 0) continue;     
                if (!isPossible(i, j)) continue;
                if (i == j) ans += age[i] * (age[i] - 1);
                else ans += age[i] * age[j];
            }
        }
        return ans;
    }

    static boolean isPossible(int x, int y) {
        if (y <= 0.5 * x + 7) return false;
        if (y > x) return false;
        if (y > 100 && x < 100) return false;
        return true;
    }
}